distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+                raise RuntimeError("No active INET socket found?")
+
+            sock = get_socket(self._servers[0])
+            self.bound_address = sock.getsockname()[:2]
+        return self.bound_address
+
+    @property
+    def listen_address(self):
+        """
+        The listening address as a string.
+        """
+        return self.prefix + unparse_host_port(*self.get_host_port())
+
+    @property
+    def contact_address(self):
+        """
+        The contact address as a string.
+        """
+        host, port = self.get_host_port()
+        host = ensure_concrete_host(host, default_host=self.default_host)
+        return self.prefix + unparse_host_port(host, port)
+
+
+class TLSListener(TCPListener):
+    prefix = "tls://"
+    comm_class = TLS
+
+    def _get_extra_kwargs(self, address, **kwargs):
+        ctx = _expect_tls_context(kwargs)
+        return {"ssl": ctx}
+
+
+class TCPBackend(Backend):
+    _connector_class = TCPConnector
+    _listener_class = TCPListener
+
+    def get_connector(self):
+        return self._connector_class()
+
+    def get_listener(self, loc, handle_comm, deserialize, **connection_args):
+        return self._listener_class(loc, handle_comm, deserialize, **connection_args)
+
+    def get_address_host(self, loc):
+        return parse_host_port(loc)[0]
+
+    def get_address_host_port(self, loc):
+        return parse_host_port(loc)
+
+    def resolve_address(self, loc):
+        host, port = parse_host_port(loc)
+        return unparse_host_port(ensure_ip(host), port)
+
+    def get_local_address_for(self, loc):
+        host, port = parse_host_port(loc)
+        host = ensure_ip(host)
+        if ":" in host:
+            local_host = get_ipv6(host)
+        else:
+            local_host = get_ip(host)
+        return unparse_host_port(local_host, None)
+
+
+class TLSBackend(TCPBackend):
+    _connector_class = TLSConnector
+    _listener_class = TLSListener
+
+
+# This class is based on parts of `asyncio.selector_events._SelectorSocketTransport`
+# (https://github.com/python/cpython/blob/dc4a212bd305831cb4b187a2e0cc82666fcb15ca/Lib/asyncio/selector_events.py#L757).
+class _ZeroCopyWriter:
+    """The builtin socket transport in asyncio makes a bunch of copies, which
+    can make sending large amounts of data much slower. This hacks around that.
+
+    Note that this workaround isn't used with the windows ProactorEventLoop or
+    uvloop."""
+
+    # We use sendmsg for scatter IO if it's available. Since bookkeeping
+    # scatter IO has a small cost, we want to minimize the amount of processing
+    # we do for each send call. We assume the system send buffer is < 4 MiB
+    # (which would be very large), and set a limit on the number of buffers to
+    # pass to sendmsg.
+    if hasattr(socket.socket, "sendmsg"):
+        try:
+            SENDMSG_MAX_COUNT = os.sysconf("SC_IOV_MAX")
+        except Exception:
+            SENDMSG_MAX_COUNT = 16  # Should be supported on all systems
```

Comment:
Is the highest value supported on all systems? If so, which system has this limit? Maybe worth including in the comment if we ever revisit

Reply:
This is the minimum value required by posix, and is what's supported on older (much older) versions of linux.